src/libraries/Microsoft.PowerFx.Connectors/PagesRecordValueAndTable.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Connectors
+{
+    // PagedRecordValue is a RecordValue that can be paged through.
+    // It is composed of the "nextLink" string and a table with the results
+    internal class PagedRecordValue : RecordValue
+    {
+        public readonly int MaxRows;
+        public readonly RecordValue CurrentPage;
+        public readonly TableValue CurrentTable;
+        public readonly string TableFieldName;
+        public readonly CancellationToken CancellationToken;
+
+        // Method used to retrieve the next page
+        public readonly Func<Task<FormulaValue>> GetNextRecordAsync;
+
+        public PagedRecordValue(RecordValue recordValue, Func<Task<FormulaValue>> getNextRecordAsync, int maxRows, CancellationToken cancellation)
+            : base(recordValue.Type)
+        {
+            CurrentPage = recordValue;
+            GetNextRecordAsync = getNextRecordAsync;
+            MaxRows = maxRows;
+            CancellationToken = cancellation;
+
+            NamedValue tableProp = CurrentPage.Fields.FirstOrDefault((NamedValue nv) => nv.Value is TableValue) ?? throw new InvalidOperationException("PagedRecordValue must contain a table");
+            TableFieldName = tableProp.Name;
+            CurrentTable = tableProp.Value as TableValue;
+        }
+
+        protected override bool TryGetField(FormulaType fieldType, string fieldName, out FormulaValue result)
+        {
+            bool b;
+            (b, result) = CurrentPage.TryGetFieldAsync(fieldType, fieldName, CancellationToken.None).ConfigureAwait(false).GetAwaiter().GetResult();
+
+            if (fieldName == TableFieldName)
+            {
+                result = new PagedTableValue(this);
+            }
+
+            return b;
+        }
+    }
+
+    internal class PagedTableValue : TableValue
```

Comment:
>PagedTableValue [](http://example.com/codeflow?start=19&length=15)

nit: 
PagesREcordValueAndTable.cs is  a unusual naming convention. 

We could just put PagedTableValue as a private nested class in PagedRecordValue (since that's the only place that should create it) and then call this file PagedRecordValue.cs